native/dll/Win32DLL/SageTVWin32DLL.cpp

```diff
@@ -1013,7 +1011,7 @@ void loadAWTLib()
 		*goodSlash = 0;
 		goodSlash = strrchr(jvmPath, '\\');
 		if (!goodSlash) return;
-		strcpy(goodSlash + 1, "jawt.dll");
+		strcpy_s(goodSlash + 1, sizeof(goodSlash) + 1, "jawt.dll");

```

Comment:
You can't use sizeof(goodslash) because it'll give you the pointer size. What you can do instead is:
*(goodSlash +1) = 0;
strncat_s(jvmPath, sizeof(jvmPath), "jawt.dll");

Reply:
I'm struggling with this one.  I don't understand how it works now.
We get a ptr with the last \ (Understand that)
We find the 2nd to last  \ (Don't understand)
How do we end up with the required resultant string in jvmPath?
initial jvmPath = C:\Program Files\Java\jre1.8.0_241\bin\server\jvm.dll
jvmPath in sageLoadedAwtLib = LoadLibrary(jvmPath); = C:\Program Files\Java\jre1.8.0_241\bin\jawt.dll

Reply:
Very slowly getting there `This pointer (plus one) will point to the base name of the file`
Still trying to figure out how that becomes the new value on jvmPath

Reply:
OK, so here's what the current code does:

1. Find the last backslash and get the pointer to it:
char* goodSlash = strrchr(jvmPath, '\\');
2. If we didn't find it, return.
if (!goodSlash) return;
3. Replace the character at the last slash with the null terminator. The essentially truncates the string so the last character is now the one before the original last slash (i.e. it removes \jvm.dll)
*goodSlash = 0;
4. Find the last slash in the string again. This will be the second one originally, but now that is the last one.
goodSlash = strrchr(jvmPath, '\\');

Then at that point we can copy into that position after the slash (original code). Or what I'm suggesting is to then truncate that string again after that slash (i.e. it would end with bin\) and then do a strcat_s on it since we know the size of jvmPath easily. (you could also do other pointer math to get a proper value for strcpy_s, but that's easier to get wrong)

Make sense now? :)

Reply:
Thanks for spending the time walking me through this 👍 it's making more sense. When it it's doing this, is it changing jvmPath too?

Reply:
Thanks for spending the time walking me through this 👍 it's making more sense. When it it's doing this, is it changing jvmPath too?

Reply:
Yes, it's changing jvmPath. When that first strrchr is called, it's setting goodSlash to be a pointer into the jvmPath array. So if you modify goodSlash, you are actually modifying jvmPath since it's pointing inside of its array.

Reply:
Ahhh. Ok. I'll finish those changes when I get home
Thanks once more :)